conjurapi/authn_test.go

```diff
@@ -1,92 +1,114 @@
-		})
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				// SETUP
+				config := &Config{}
+				config.mergeEnv()
 
-		Convey("Rotate the API key of a foreign role of non-user kind", func() {
-			bobAPIKey, err := conjur.RotateAPIKey("cucumber:host:bob")
+				apiKey := os.Getenv("CONJUR_AUTHN_API_KEY")
+				login := os.Getenv("CONJUR_AUTHN_LOGIN")
 
-			_, err = conjur.Authenticate(authn.LoginPair{"host/bob", string(bobAPIKey)})
-			So(err, ShouldBeNil)
-		})
+				policy := `
+- !user alice
+- !host bob
+`
 
-		Convey("Rotate the API key of a foreign role and read the data stream", func() {
-			rotateResponse, err := conjur.RotateAPIKeyReader("cucumber:user:alice")
+				conjur, err := NewClientFromKey(*config, authn.LoginPair{Login: login, APIKey: apiKey})
+				assert.NoError(t, err)
 
-			So(err, ShouldBeNil)
-			aliceAPIKey, err := ReadResponseBody(rotateResponse)
-			So(err, ShouldBeNil)
+				conjur.LoadPolicy(
+					PolicyModePut,
+					"root",
+					strings.NewReader(policy),
+				)
 
-			_, err = conjur.Authenticate(authn.LoginPair{"alice", string(aliceAPIKey)})
-			So(err, ShouldBeNil)
-		})
+				// EXERCISE
+				runAssertions(t, tc, config, login, apiKey)
+			})
+		}
 	})
 
 	if os.Getenv("TEST_VERSION") != "oss" {
-		Convey("V4", t, func() {
-
-			config := &Config{
-				ApplianceURL: os.Getenv("CONJUR_V4_APPLIANCE_URL"),
-				SSLCert:      os.Getenv("CONJUR_V4_SSL_CERTIFICATE"),
-				Account:      os.Getenv("CONJUR_V4_ACCOUNT"),
-				V4:           true,
-			}
+		t.Run("V4", func(t *testing.T) {
 
-			login := os.Getenv("CONJUR_V4_AUTHN_LOGIN")
-			apiKey := os.Getenv("CONJUR_V4_AUTHN_API_KEY")
+			for _, tc := range testCases {
+				t.Run(tc.name, func(t *testing.T) {
+					// SETUP
+					config := &Config{
+						ApplianceURL: os.Getenv("CONJUR_V4_APPLIANCE_URL"),
+						SSLCert:      os.Getenv("CONJUR_V4_SSL_CERTIFICATE"),
+						Account:      os.Getenv("CONJUR_V4_ACCOUNT"),
+						V4:           true,
+					}
 
-			conjur, err := NewClientFromKey(*config, authn.LoginPair{login, apiKey})
-			So(err, ShouldBeNil)
-
-			Convey("Rotate the API key of a foreign role of kind user", func() {
-				aliceAPIKey, err := conjur.RotateAPIKey("cucumber:user:alice")
-
-				_, err = conjur.Authenticate(authn.LoginPair{"alice", string(aliceAPIKey)})
-				So(err, ShouldBeNil)
-			})
+					login := os.Getenv("CONJUR_V4_AUTHN_LOGIN")
+					apiKey := os.Getenv("CONJUR_V4_AUTHN_API_KEY")
 
-			Convey("Rotate the API key of a foreign role of non-user kind", func() {
-				bobAPIKey, err := conjur.RotateAPIKey("cucumber:host:bob")
+					// EXERCISE
+					runAssertions(t, tc, config, login, apiKey)
+				})
 
-				_, err = conjur.Authenticate(authn.LoginPair{"host/bob", string(bobAPIKey)})
-				So(err, ShouldBeNil)
-			})
+			}
 		})
 	}
 }
+
+func runAssertions(t *testing.T, tc testCase, config *Config, login string, apiKey string) {
+	conjur, err := NewClientFromKey(*config, authn.LoginPair{Login: login, APIKey: apiKey})
```

Comment:
Oooh, these 2 lines:
```
	conjur, err := NewClientFromKey(*config, authn.LoginPair{Login: login, APIKey: apiKey})
	assert.NoError(t, err)
```
also do not depend upon the test cases, and can be moved up to the common `SETUP` sections that can be split out for V5 and V4.

If these 2 lines are moved out of the `runAssertions()` function, then you wouldn't have to pass in these arguments:
```
config *Config, login string, apiKey string
```
to runAssertions().